src/codecoverage/backend/codecoverage_backend/services/gcp.py

```diff
@@ -0,0 +1,248 @@
+
+        for page in range(nb_pages):
+
+            r = requests.get(HGMO_PUSHES_URL.format(repository=repository), params=params)
+            data = r.json()
+
+            # Sort pushes to go from oldest to newest
+            pushes = sorted([
+                (int(push_id), push)
+                for push_id, push in data['pushes'].items()
+            ], key=lambda p: p[0])
+            if not pushes:
+                return
+
+            for push_id, push in pushes:
+
+                changesets = push['changesets']
+                self.store_push(repository, push_id, changesets)
+
+                reports = [
+                    changeset
+                    for changeset in changesets
+                    if self.ingest_report(repository, push_id, changeset)
+                ]
+                if reports:
+                    logger.info('Found reports in that push', push_id=push_id)
+
+            newest = pushes[-1][0]
+            params['startID'] = newest
+            params['endID'] = newest + chunk_size
+
+    def ingest_report(self, repository, push_id, changeset):
+        '''
+        When a report exist for a changeset, download it and update redis data
+        '''
+        assert isinstance(push_id, int)
+
+        # Download the report
+        if not self.download_report(repository, changeset):
+            return
+
+        # Add the changeset to the sorted sets of known reports
+        # The numeric push_id is used as a score to keep the ingested
+        # changesets ordered
+        self.redis.zadd(KEY_REPORTS.format(repository=repository), {changeset: push_id})
+
+        logger.info('Ingested report', changeset=changeset)
+        return True
+
+    def download_report(self, repository, changeset):
+        '''
+        Download and extract a json+zstd covdir report
+        '''
+        # Chek the report is available on remote storage
+        path = '{}/{}.json.zstd'.format(repository, changeset)
+        blob = self.bucket.blob(path)
+        if not blob.exists():
+            logger.debug('No report found on GCP', path=path)
+            return False
+
+        archive_path = os.path.join(self.reports_dir, blob.name)
+        json_path = os.path.join(self.reports_dir, blob.name.rstrip('.zstd'))
+        if os.path.exists(json_path):
+            logger.info('Report already available', path=json_path)
+            return True
+
+        os.makedirs(os.path.dirname(archive_path), exist_ok=True)
+        blob.download_to_filename(archive_path)
+        logger.info('Downloaded report archive', path=archive_path)
+
+        with open(json_path, 'wb') as output:
+            with open(archive_path, 'rb') as archive:
+                dctx = zstd.ZstdDecompressor()
+                reader = dctx.stream_reader(archive)
+                while True:
+                    chunk = reader.read(16384)
+                    if not chunk:
+                        break
+                    output.write(chunk)
+
```

Comment:
Should we remove the zstd file now?

Reply:
As long as we are on heroku yes.
When we switch to GCP, we will have some FS persistence, and could retain archives as long term storage (and only keep the warm json extracted)